src/lib/prov/tpm2/tpm2_crypto_backend/tpm2_crypto_backend.cpp

```diff
@@ -0,0 +1,859 @@
+      *out_size = encrypted.size();
+      Botan::copy_mem(std::span{out_buffer, *out_size}, encrypted);
+
+      return TSS2_RC_SUCCESS;
+   });
+   #else
+   BOTAN_UNUSED(pub_tpm_key, in_size, in_buffer, max_out_size, out_buffer, out_size, label, userdata);
+   return TSS2_ESYS_RC_NOT_IMPLEMENTED;
+   #endif
+}
+
+/** Computation of an ephemeral ECC key and shared secret Z.
+ *
+ * According to the description in TPM spec part 1 C 6.1 a shared secret
+ * between application and TPM is computed (ECDH). An ephemeral ECC key and a
+ * TPM key are used for the ECDH key exchange.
+ * @param[in] key The key to be used for ECDH key exchange.
+ * @param[in] max_out_size the max size for the output of the public key of the
+ *            computed ephemeral key.
+ * @param[out] Z The computed shared secret.
+ * @param[out] Q The public part of the ephemeral key in TPM format.
+ * @param[out] out_buffer The public part of the ephemeral key will be marshaled
+ *             to this buffer.
+ * @param[out] out_size The size of the marshaled output.
+ * @param[in,out] userdata information.
+ * @retval TSS2_RC_SUCCESS on success
+ * @retval USER_DEFINED user defined errors on failure.
+ */
+static TSS2_RC get_ecdh_point(TPM2B_PUBLIC* key,
+                              size_t max_out_size,
+                              TPM2B_ECC_PARAMETER* Z,
+                              TPMS_ECC_POINT* Q,
+                              BYTE* out_buffer,
+                              size_t* out_size,
+                              void* userdata) {
+   #if defined(BOTAN_HAS_TPM2_ECC_ADAPTER)
+   return thunk([&] {
+      BOTAN_UNUSED(userdata);
+      BOTAN_ASSERT_NONNULL(key);
+      BOTAN_ASSERT_NOMSG(key->publicArea.type == TPM2_ALG_ECC);
+
+      const auto curve_name = Botan::TPM2::curve_id_tss2_to_botan(key->publicArea.parameters.eccDetail.curveID);
+      const auto curve_order_byte_size =
+         Botan::TPM2::curve_id_order_byte_size(key->publicArea.parameters.eccDetail.curveID);
+      if(!curve_name || !curve_order_byte_size) {
+         return TSS2_ESYS_RC_NOT_IMPLEMENTED;  // ECC curve not supported
+      }
+
+      const auto domain = Botan::EC_Group::from_name(curve_name.value());
+
+      BOTAN_ASSERT_NOMSG(curve_order_byte_size.value() == domain.get_p_bytes());
+
+      Botan::AutoSeeded_RNG rng;  //TODO: What RNG to use?
+
+      // 1: Generate ephemeral key
+      const Botan::ECDH_PrivateKey eph_key(rng, domain);
+
+      // Get public key coordinates
+      const Botan::PointGFp& eph_pub_point = eph_key.public_point();
+      const auto& bn_x = eph_pub_point.get_affine_x();
+      const auto& bn_y = eph_pub_point.get_affine_y();
+
+      // Write public ephemeral affine coordinates to TPMS_ECC_POINT (Q)
+      if(bn_x.bytes() != curve_order_byte_size || bn_y.bytes() != curve_order_byte_size) {
+         // TODO: Is actually correct to check bytes() > curve_order_byte_size?
+         return TSS2_ESYS_RC_GENERAL_FAILURE;
+      }
```

Comment:
As far as I remember, `BigInt::bytes()` **does not** take zero-bytes into account. So if the coordinate has a zero byte by chance, `<=` might actually be better here, as you probably imply in the TODO.

Reply:
This check is actually performed in `BigInt::serialize(curve_order_byte_size)` so I removed it